storage/src/qmdb/any/ordered/mod.rs

```diff
@@ -1078,13 +1081,231 @@ impl<
+                // mutations set. Any colliding key could potentially be a previous key of some new
+                // or deleted key.
+                possible_previous.insert(key, (key_data.value, old_loc));
+                continue;
+            };
+
+            if let Some(value) = update {
+                // This is an update of an existing key.
+                updated.insert(key.clone(), (value, old_loc));
+                possible_next.insert(key_data.next_key.clone());
+            } else {
+                // This is a delete of an existing key.
+                deleted.push(key.clone());
+
+                // The next_key of a deleted key will need to be the next_key of some other remaining key, unless it too
+                // ends up deleted.
+                possible_next.insert(key_data.next_key.clone());
+
+                // Update the log and snapshot.
+                delete_known_loc(&mut self.snapshot, &key, old_loc);
+                self.log.append(C::Item::new_delete(key.clone())).await?;
+
+                // Each delete reduces the active key count by one and inactivates that key.
+                self.active_keys -= 1;
+                self.steps += 1;
+            }
+
+            // Any key could collide with some other key, so we must record each as a "possible previous".
+            possible_previous.insert(key, (key_data.value, old_loc));
+        }
+
+        // Any key remaining in `mutations` must be a new key so move it to the created map.
+        for (key, value) in mutations {
+            let Some(value) = value else {
+                continue; // can happen from attempt to delete a non-existent key
+            };
+
+            created.insert(key.clone(), value);
+
+            // Any newly created key must be a next_key for some remaining key.
+            possible_next.insert(key);
+        }
+
+        // Complete the `possible_previous` and `possible_next` sets by including entries from the
+        // previous _translated_ key for any created or deleted key.
+        let mut locations = Vec::new();
+        for key in deleted.iter().chain(created.keys()) {
+            let iter = self.snapshot.prev_translated_key(key);
+            locations.extend(iter.copied());
+        }
+        locations.sort();
+        locations.dedup();
+        let futures = locations.iter().map(|loc| self.log.read(*loc));
+        let results = try_join_all(futures).await?;
+
+        for (op, old_loc) in (results.into_iter()).zip(locations) {
+            let key = op.key().expect("updates should have a key").clone();
+            let key_data = op.into_key_data().expect("updates should have key data");
+            possible_next.insert(key_data.next_key);
+            possible_previous.insert(key, (key_data.value, old_loc));
+        }
+
+        // Remove deleted keys from the possible_* sets.
+        for key in deleted.iter() {
+            possible_previous.remove(key);
+            possible_next.remove(key);
```

Comment:
In the loop above we're adding the `deleted` elements into `possible_*` only to remove them here

Reply:
The loop above is inserting the "previous translated keys" to each deleted element, not the deleted elements themselves.